docs/architecture/grpc-for-wcf-developers/rpc-types.md

```diff
@@ -0,0 +1,268 @@
+
+### gRPC streaming
+
+gRPC supports the creation of persistent streams from client to server, and from server to client. Both types of stream may be active concurrently; this is called bidirectional streaming. Streams can be used for arbitrary, asynchronous messaging over time, or for passing large datasets that are too big to generate and send in a single request or response.
+
+The following example shows a server streaming RPC.
+
+```protobuf
+service ClockStreamer {
+    rpc Subscribe(ClockSubscribeRequest) returns (stream ClockMessage);
+}
+```
+
+```csharp
+public class ClockStreamerService : ClockStreamer.ClockStreamerBase
+{
+    public override async Task Subscribe(ClockSubscribeRequest request,
+        IServerStreamWriter<ClockMessage> responseStream,
+        ServerCallContext context)
+    {
+        while (!context.CancellationToken.IsCancellationRequested)
+        {
+            var time = DateTimeOffset.UtcNow;
+            await responseStream.WriteAsync(new ClockMessage { message = $"The time is {time:t}." });
+            await Task.Delay(TimeSpan.FromSeconds(10), context.CancellationToken);
+        }
+    }
+}
+```
+
+This server stream could be consumed from a client application as shown in the following code:
+
+```csharp
+public async Task TellTheTimeAsync(CancellationToken token)
+{
+    var channel = GrpcChannel.ForAddress("https://localhost:5001");
+    var client = new ClockStreamer.ClockStreamerClient(channel);
+
+    var request = new ClockSubscribeRequest();
+    var response = client.Subscribe(request);
+
+    await foreach (var update in response.ResponseStream.ReadAllAsync(token))
+    {
+        Console.WriteLine(update.Message);
+    }
+}
+```
+
+> [!NOTE]
+> Server streaming RPCs are useful for subscription-style services, and also for sending very large datasets when it would be inefficient or impossible to build the entire dataset in memory. However, streaming responses isn't as fast as sending `repeated` fields in a single message, so as a rule streaming shouldn't be used for small datasets.
+
+### Differences to WCF
+
+A WCF duplex service uses a client callback interface that can have multiple methods. A gRPC server streaming service can only send messages over a single stream. If you need multiple methods, use a message type with either [an Any field or a oneof field](protobuf-any-oneof.md) to send different messages, and write code in the client to handle them.
+
+In WCF, the one-way method called from the client returns immediately, but the [ServiceContract](https://docs.microsoft.com/dotnet/api/system.servicemodel.servicecontractattribute?view=netframework-4.8) class with the session is kept alive until the connection is closed. In gRPC, the `Task` returned by the implementation method shouldn't complete until the connection is closed.
```

Comment:
It doesn't return immediately as the server can still return an error. For example if you send a message for an action that the server doesn't recognize, a fault will be returned to the client and the client will throw an exception. So the server will verify that the message is for an operation it can dispatch and that the soap message is valid. For HTTP, the server replies with a 200 OK response with a 0 length content, for framing protocols such as NetTcp, it returns a single byte ACK. Only then does the client complete the Task/return from sync call.